modules/_labs/combobox/react/lib/Combobox.tsx

```diff
@@ -0,0 +1,422 @@
+  /**
+   * Initial value to set the input to
+   */
+  initialValue?: string;
+  /**
+   * The type of icon button to use for clearing input
+   */
+  clearButtonVariant: IconButtonVariant;
+  /**
+   * Show button to clear input field
+   */
+  showClearButton?: boolean;
+  /**
+   * Aria Label for clear button
+   */
+  clearButtonLabel: string;
+  /**
+   * An array of menu items to show under the TextInput
+   */
+  autocompleteItems?: React.ReactElement<MenuItemProps>[];
+  /**
+   * Callback to listen when the TextInput changes
+   */
+  onChange?: React.ChangeEventHandler<HTMLInputElement>;
+  /**
+   * Callback to listen when the TextInput focuses
+   */
+  onFocus?: React.FocusEventHandler;
+  /**
+   * Callback to listen when the TextInput blurs
+   */
+  onBlur?: React.FocusEventHandler;
+  /**
+   * The id of the form field.
+   */
+  labelId?: string;
+}
+
+export interface ComboboxState {
+  value: string;
+  isFocused: boolean;
+  showingAutocomplete: boolean;
+  selectedAutocompleteIndex: number | null;
+}
+
+const Container = styled('div')<Pick<ComboboxProps, 'grow'>>(
+  {
+    display: 'inline-block',
+  },
+  ({grow}) => ({
+    width: grow ? '100%' : 'auto',
+  })
+);
+
+const Status = styled('div')({
+  ...accessibleHide,
+});
+
+const InputContainer = styled('div')({
+  display: `flex`,
+  alignItems: `center`,
+  position: 'relative',
+});
+
+const MenuContainer = styled(Card)({
+  position: 'absolute',
+  zIndex: 1,
+  left: 0,
+  top: '100%',
+  borderRadius: borderRadius.m,
+  background: commonColors.background,
+  border: `none`,
+  marginTop: `-${borderRadius.m}`,
+  width: '100%',
+  minWidth: 0,
+});
+
+const AutocompleteList = styled('ul')({
+  padding: 0,
+  margin: `${spacing.xxs} 0`,
+});
+
+const ResetButton = styled(IconButton)<{shouldShow: boolean}>(
+  {
+    width: spacing.l,
+    height: spacing.l,
+    margin: `-${spacing.s}`,
+    left: `-${spacing.m}`,
+    padding: 0,
+    zIndex: 2,
+    transition: 'opacity 120ms',
+  },
+  ({shouldShow}) => ({
+    visibility: shouldShow ? 'visible' : 'hidden',
+    opacity: shouldShow ? 1 : 0,
+  })
+);
+
+const listBoxIdPart = `listbox`;
+const optionIdPart = `option`;
+const getOptionId = (baseId?: string, index?: number) => `${baseId}-${optionIdPart}-${index}`;
+
+export default class Combobox extends React.Component<ComboboxProps, ComboboxState> {
+  private inputRef = this.props.children.props.inputRef || React.createRef<HTMLInputElement>();
+  private comboboxRef = React.createRef<HTMLDivElement>();
+  private id = uuid();
+
+  static defaultProps = {
+    clearButtonLabel: `Reset Search Input`,
+    clearButtonVariant: IconButton.Variant.Plain,
+  };
+
+  state: Readonly<ComboboxState> = {
+    isFocused: false,
+    value: '',
+    showingAutocomplete: false,
+    selectedAutocompleteIndex: null,
+  };
+
+  componentDidMount() {
+    if (this.props.initialValue) {
+      this.setInputValue(this.props.initialValue);
```

Comment:
This seems familiar to me. I think I've asked this before - why do we need to fire an event when the component is initialized with a value? Would this cause problems if an event handler is expecting to change only based on user interactions? They'd have to do some crazy logic to ignore changes the first time the component is rendered.

Reply:
Without it an initialValue won't be able to fetch autocomplete items